src/CachePeer.cc

```diff
@@ -47,6 +48,56 @@ CachePeer::~CachePeer()
     xfree(domain);
 }
 
+void
+CachePeer::noteSuccess()
+{
+    if (!tcp_up) {
+        debugs(15, 2, "TCP connection to " << host << "/" << http_port << " succeeded");
+        tcp_up = connect_fail_limit; // NP: so peerAlive() works properly.
+        peerAlive(this);
+    } else {
+        tcp_up = connect_fail_limit;
+    }
+}
+
+void
+CachePeer::noteFailure(const Http::StatusCode code)
+{
+    if (Http::Is4xx(code))
+        return; // this failure is not our fault
+
+    countFailure();
+}
+
+// TODO: Require callers to detail failures instead of using one (and often
+// misleading!) "TCP" label for all of them.
+/// noteFailure() helper for handling failures attributed to this peer
+void
+CachePeer::countFailure()
+{
+    stats.last_connect_failure = squid_curtime;
+    if (tcp_up > 0)
+        --tcp_up;
+
+    // TODO: Report peer name. Same-addresses peers often have different names.
+
+    const auto consideredAliveByAdmin = stats.logged_state == PEER_ALIVE;
+    const auto level = consideredAliveByAdmin ? DBG_IMPORTANT : 2;
+    debugs(15, level, "ERROR: TCP connection to " << host << "/" << http_port << " failed");
```

Comment:
Let's not change this TCP-based wording in _this_ PR. We have another PR that polishes cache_peer reporting (already in the official review) that will modify this line and another in-progress project that will actually properly detail these failures. This PR is not a good opportunity to fix the misleading TCP label.

Said that, if you insist, I will commit your suggestion to avoid blocking this PR.

Reply:
Huh? the line is being touched by this PR, there is no better time to fix a bug like this.

Reply:
> Huh? the line is being touched by this PR, there is no better time to fix a bug like this.

You are thinking that removing TCP fixes a bug. I believe it just obscures the real problem, and the change itself is out-of-scope.

Removed in 7aa16d6 in hope to unblock this PR.